### ISSUE TITLE
Graceful node drain results in never-cleaned-up pods in various Error/Completed phases

### ISSUE THREAD

**Issue by buroa** [kubernetes/kubernetes#122122]
### What happened?

During a graceful node drain, I am seeing various pods in error / completed pod phases that are never cleaned up (deleted). They stay until I manually delete them with either a controller like https://github.com/kubernetes-sigs/descheduler or `kubectl delete`.

`postgres-1` here is just a pod with no statefulset / deployment. It was created manually by cloudnative-pg.

```
cloudnative-pg-74c458f8cf-257g9   0/1     Completed   0          10h
cloudnative-pg-74c458f8cf-zdb2k   1/1     Running     0          7m46s
postgres-1                        0/1     Error       0          10h
postgres-2                        1/1     Running     0          13h
postgres-3                        1/1     Running     0          14h
redis-node-0                      3/3     Running     0          10h
redis-node-1                      3/3     Running     0          13h
redis-node-2                      3/3     Running     0          14h
```

```
Name:             postgres-1
Namespace:        databases
Priority:         0
Service Account:  postgres
Node:             w2/10.0.0.22
Start Time:       Wed, 29 Nov 2023 18:45:01 -0600
Labels:           cnpg.io/cluster=postgres
                  cnpg.io/instanceName=postgres-1
                  cnpg.io/instanceRole=replica
                  cnpg.io/podRole=instance
                  role=replica
Annotations:      cnpg.io/nodeSerial: 1
                  cnpg.io/operatorVersion: 1.21.1
                  cnpg.io/podEnvHash: 8566455577
                  cnpg.io/podSpec:
                    {"volumes":[{"name":"pgdata","persistentVolumeClaim":{"claimName":"postgres-1"}},{"name":"scratch-data","emptyDir":{}},{"name":"shm","empt...
                  k8s.v1.cni.cncf.io/network-status:
                    [{
                        "name": "cilium",
                        "interface": "eth0",
                        "ips": [
                            "10.244.5.94"
                        ],
                        "mac": "0e:ea:e8:5c:2c:5f",
                        "default": true,
                        "dns": {},
                        "gateway": [
                            "10.244.5.192"
                        ]
                    }]
                  kyverno.io/exclude: true
Status:           Failed
Reason:           Terminated
Message:          Pod was terminated in response to imminent node shutdown.
SeccompProfile:   RuntimeDefault
IP:
IPs:              <none>
Controlled By:    Cluster/postgres
Init Containers:
  bootstrap-controller:
    Container ID:    containerd://d19c819b1d25fbed28030d500672fdf9cacee69e61dfd5c1fa8cec6e744976d9
    Image:           ghcr.io/cloudnative-pg/cloudnative-pg:1.21.1
    Image ID:        ghcr.io/cloudnative-pg/cloudnative-pg@sha256:9f707d91de1c92975ab1a098b52e804d8f5b90c7934a0b1c40d00e773522a123
    Port:            <none>
    Host Port:       <none>
    SeccompProfile:  RuntimeDefault
    Command:
      /manager
      bootstrap
      /controller/manager
      --log-level=info
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 29 Nov 2023 18:45:04 -0600
      Finished:     Wed, 29 Nov 2023 18:45:04 -0600
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /controller from scratch-data (rw)
      /dev/shm from shm (rw)
      /etc/app-secret from app-secret (rw)
      /etc/superuser-secret from superuser-secret (rw)
      /run from scratch-data (rw)
      /var/lib/postgresql/data from pgdata (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-q7928 (ro)
Containers:
  postgres:
    Container ID:    containerd://ba3e10ebb1f57e8ed4da10800ec954ea16924105f483ff3b099874cab157c722
    Image:           ghcr.io/cloudnative-pg/postgresql:16.1-4
    Image ID:        sha256:189647f611df17d25679d38dc453c474299bd65cebc6f14635054d471b13e9cd
    Ports:           5432/TCP, 9187/TCP, 8000/TCP
    Host Ports:      0/TCP, 0/TCP, 0/TCP
    SeccompProfile:  RuntimeDefault
    Command:
      /controller/manager
      instance
      run
      --log-level=info
    State:          Terminated
      Reason:       Error
      Exit Code:    137
      Started:      Wed, 29 Nov 2023 18:45:05 -0600
      Finished:     Thu, 30 Nov 2023 05:37:00 -0600
    Ready:          False
    Restart Count:  0
    Liveness:       http-get http://:8000/healthz delay=0s timeout=5s period=10s #success=1 #failure=3
    Readiness:      http-get http://:8000/readyz delay=0s timeout=5s period=10s #success=1 #failure=3
    Startup:        http-get http://:8000/healthz delay=0s timeout=5s period=10s #success=1 #failure=360
    Environment:
      PGDATA:        /var/lib/postgresql/data/pgdata
      POD_NAME:      postgres-1
      NAMESPACE:     databases
      CLUSTER_NAME:  postgres
      PGPORT:        5432
      PGHOST:        /controller/run
    Mounts:
      /controller from scratch-data (rw)
      /dev/shm from shm (rw)
      /etc/app-secret from app-secret (rw)
      /etc/superuser-secret from superuser-secret (rw)
      /run from scratch-data (rw)
      /var/lib/postgresql/data from pgdata (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-q7928 (ro)
Conditions:
  Type               Status
  DisruptionTarget   True
  Initialized        True
  Ready              False
  ContainersReady    False
  PodScheduled       True
Volumes:
  pgdata:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  postgres-1
    ReadOnly:   false
  scratch-data:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  shm:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  superuser-secret:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  cloudnative-pg-secret
    Optional:    false
  app-secret:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  postgres-app
    Optional:    false
  kube-api-access-q7928:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type    Reason   Age   From     Message
  ----    ------   ----  ----     -------
  Normal  Killing  10m   kubelet  Stopping container postgres
  ```

Various logging pods:

```
loki-write-0                                                1/1     Running     0             14h     10.244.4.251   w0     <none>           <none>
loki-write-1                                                0/1     Completed   0             11h     <none>         w2     <none>           <none>
loki-write-2                                                1/1     Running     0             14h     10.244.3.29    w1     <none>           <none>
```

```
Name:             loki-write-1
Namespace:        monitoring
Priority:         0
Service Account:  loki
Node:             w2/10.0.0.22
Start Time:       Wed, 29 Nov 2023 18:42:31 -0600
Labels:           app.kubernetes.io/component=write
                  app.kubernetes.io/instance=loki
                  app.kubernetes.io/name=loki
                  app.kubernetes.io/part-of=memberlist
                  apps.kubernetes.io/pod-index=1
                  controller-revision-hash=loki-write-7ff4f96bf9
                  statefulset.kubernetes.io/pod-name=loki-write-1
Annotations:      checksum/config: 67193d214dd0992612d9cefaa5d8f0ca0fc148cbc6943ca6cd8f8aa75d2b9057
                  k8s.v1.cni.cncf.io/network-status:
                    [{
                        "name": "cilium",
                        "interface": "eth0",
                        "ips": [
                            "10.244.5.245"
                        ],
                        "mac": "ea:47:d8:cc:b7:e1",
                        "default": true,
                        "dns": {},
                        "gateway": [
                            "10.244.5.192"
                        ]
                    }]
                  kubectl.kubernetes.io/restartedAt: 2023-11-21T20:33:18-06:00
                  kyverno.io/exclude: true
Status:           Succeeded
IP:
IPs:              <none>
Controlled By:    StatefulSet/loki-write
Containers:
  loki:
    Container ID:  containerd://8705a1d6d2c36fa4c98c1d8f281f2146ae2e410462df7a9cfe92d373421c7cb1
    Image:         docker.io/grafana/loki:2.9.2
    Image ID:      sha256:314b95a20590e79e06a9d244a95ee7320f8762141b446f40f8d74e1e887f5494
    Ports:         3100/TCP, 9095/TCP, 7946/TCP
    Host Ports:    0/TCP, 0/TCP, 0/TCP
    Args:
      -config.file=/etc/loki/config/config.yaml
      -target=write
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 29 Nov 2023 18:42:35 -0600
      Finished:     Thu, 30 Nov 2023 05:36:40 -0600
    Ready:          False
    Restart Count:  0
    Readiness:      http-get http://:http-metrics/ready delay=30s timeout=1s period=10s #success=1 #failure=3
    Environment:    <none>
    Mounts:
      /etc/loki/config from config (rw)
      /etc/loki/runtime-config from runtime-config (rw)
      /var/loki from data (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-4hgg6 (ro)
Conditions:
  Type               Status
  DisruptionTarget   True
  Initialized        True
  Ready              False
  ContainersReady    False
  PodScheduled       True
Volumes:
  data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  data-loki-write-1
    ReadOnly:   false
  config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      loki
    Optional:  false
  runtime-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      loki-runtime
    Optional:  false
  kube-api-access-4hgg6:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type    Reason   Age   From     Message
  ----    ------   ----  ----     -------
  Normal  Killing  11m   kubelet  Stopping container loki
  ```

### What did you expect to happen?

The pods were cleaned up

### How can we reproduce it (as minimally and precisely as possible)?

Most likely the same as https://github.com/kubernetes/kubernetes/issues/118310#issuecomment-1698823460

Gracefully drain a node with various Statefulsets / Deployments

### Anything else we need to know?

https://github.com/kubernetes/kubernetes/issues/118310
Talos Linux v1.5.5
containerd v1.7.7
K8s v1.28.4 with feature gate GracefulNodeShutdown=true

### Kubernetes version

K8s v1.28.4

### Cloud provider

N/A

### OS version

Talos Linux v1.5.5 running Linux Kernel 6.1.61

### Install tools

N/A

### Container runtime (CRI) and version (if applicable)

containerd v1.7.7

### Related plugins (CNI, CSI, ...) and versions (if applicable)

N/A

**Comment by tvories** [comment]
I also experience this.

Talos: 1.5.4
Kubernetes: 1.28.3

**Comment by mimowo** [comment]
Since the pods are recreated I don't think this is directly related to https://github.com/kubernetes/kubernetes/issues/118310. 

It is expected behavior that kubelet initiates the termination, but does not delete the pods: https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown (at least in the Beta version of the feature). The accumulated pods would eventually be collected by PodGC, when their number exceeds 12500 (default value for `--terminated-pod-gc-threshold`). 

Still, I think this could be better handled by the Deployments controller, which could delete the terminated pods, similarly as the StatefulSet controller does [here](https://github.com/kubernetes/kubernetes/blob/55f2bc10435160619d1ece8de49a1c0f8fcdf276/pkg/controller/statefulset/stateful_set_control.go#L400), or the DaemonSet controller - which identifies terminated pods to be deleted [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemon_controller.go#L831-L862).
/sig apps
cc @alculquicondor

**Comment by alculquicondor** [comment]
I think this is WAI.
right @bobbypage?

The StatefulSet controller doesn't have another option, because it cannot create another Pod with the same name. Not sure why the Daemonset controller would delete the Pod.

In any case, deleting these Pods could upset the users that actually expect these Pods to remain to run some form of diagnostics. I don't think we can start deleting them.

I seem to remember there was an open issue to make pod garbage collection filter based on some attributes, but I can't find it.

**Comment by SergeyKanzhelev** [comment]
@JJGadgets is it graceful node termination or manually triggered drain?

Also, it may well be by design.

There is also maybe https://github.com/kubernetes/kubernetes/pull/121389 related. Would be great if you can try on later version.

/triage needs-information

Please read the comments and if this is expected behavior we should close the bug.

**Comment by JJGadgets** [comment]
@SergeyKanzhelev Graceful node termination. That's what I understand `talosctl reboot` does under the hood.

Re #121389: Looks like the fix needed for StatefulSets to me too, but I read through changelogs and commit log for the latest stable release (1.28.4) and don't see this PR being included in 1.28.4's branch.
I'm happy to update to 1.28.4 and later stable versions regardless given it's a homelab, but I don't think it will incorporate this change you want me to try.

> deleting these Pods could upset the users that actually expect these Pods to remain to run some form of diagnostics
If the node termination was initiated intentionally and completed gracefully (for example, in this case a successful `talosctl reboot` manually ran by a user), I'm not sure if said user would need to have the old pods hanging around for diagnostics, similar to how if I were to reboot my workstation, I don't expect my old shell to be retained.

When I intentionally reboot a node, I don't expect for me to have to clear pods manually for them to restart/reschedule in a timely fashion (for the deployment pods that seem to take at least 10 minutes to restart on 1.28), but instead my workloads should either reschedule to another node if possible, or be unavailable because of node placement (TSC, anti/affinity, nodeSelector), and whatever that couldn't be rescheduled to another node would then schedule onto the rebooted node once it becomes Ready. Thus, I'm not sure if I would consider this WAI.

I understand that StatefulSets are handled differently by their very nature, but similarly to waiting at least 10 minutes for deployment pods to restart/reschedule, having StatefulSet pods be completely unavailable after a reboot and stuck on "Succeeded" and requiring a user to manually clear the pod so it can reschedule feels like too much babysitting. I remember before 1.28, the only times that my StatefulSet pods required me to clear them manually was on an **un**graceful node termination, where I lost power to the node for more than an hour.

Although, having only run Kubernetes on a 1 man scale personally, I may be shortsighted on all this, and there may be changes to how the Kubernetes team views handling pods on graceful node termination, so feel free to correct and enlighten me on this so we can come to an understanding.

**Comment by mimowo** [comment]
> @SergeyKanzhelev Graceful node termination. That's what I understand `talosctl reboot` does under the hood.
> 
> Re #121389: Looks like the fix needed for StatefulSets to me too, but I read through changelogs and commit log for the latest stable release (1.28.4) and don't see this PR being included in 1.28.4's branch. I'm happy to update to 1.28.4 and later stable versions regardless given it's a homelab, but I don't think it will incorporate this change you want me to try.

The fix for StatefulSets is on the release-1.28 branch (see [here](https://github.com/kubernetes/kubernetes/commit/1058a06f256102d8425a410771cc4f838b73f3bd)), and will be released in 1.28.5.

**Comment by mimowo** [comment]
> Is this ever going to be fixed @mimowo?

TBH, I'm not sure there is anything more to be fixed.  Could you explain what your expectation is?

As pointed out [here](https://github.com/kubernetes/kubernetes/issues/122122#issuecomment-1839268455) this is WAI for the current design that graceful node shutdown leaves the terminated pods around. I will leave the decision about the design of the feature to the node team.

For some higher-level controllers, such as StatefulSet, the terminated pods need to be recreated, and to my knowledge all bugs related to this are now fixed (see [here](https://github.com/kubernetes/kubernetes/issues/122122#issuecomment-1846930001)).

Also, as pointed out [here](https://github.com/kubernetes/kubernetes/issues/122122#issuecomment-1839159729) the accumulated pods would eventually be collected by PodGC, when their number exceeds 12500 (default value for `--terminated-pod-gc-threshold`).

**Comment by onedr0p** [comment]
> Also, as pointed out https://github.com/kubernetes/kubernetes/issues/122122#issuecomment-1839159729 the accumulated pods would eventually be collected by PodGC, when their number exceeds 12500 (default value for --terminated-pod-gc-threshold)

I am on 1.29.1 deployed with Talos Linux and I don't think that is true, I had a pod from a deployment in a errored state for 15 hours and it was never cleaned up.

<img width="984" alt="image" src="https://github.com/kubernetes/kubernetes/assets/213795/664e97e1-9ea4-4b53-a52a-325ddde96e37">

**Comment by mimowo** [comment]
> > Also, as pointed out [#122122 (comment)](https://github.com/kubernetes/kubernetes/issues/122122#issuecomment-1839159729) the accumulated pods would eventually be collected by PodGC, when their number exceeds 12500 (default value for --terminated-pod-gc-threshold)
> 
> I am on 1.29.1 deployed with Talos Linux and I don't think that is true, I had pods from a deployment in a errored state for 15 hours and it was never cleaned up. This was after a successful drain / graceful shutdown of all my nodes one at a time.

The PodGC cleanup of terminated pods does not depend on time, but on the number of pods in terminal state: https://github.com/kubernetes/kubernetes/blob/50bf3a2060ea798600af1a4c125e0b62d08e9680/pkg/controller/podgc/gc_controller.go#L125-L127. The number of terminated pods needs to exceed `--terminated-pod-gc-threshold`. see here: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/

**Comment by tuibeovince** [comment]
> > Is this ever going to be fixed @mimowo?
> 
> TBH, I'm not sure there is anything more to be fixed. Could you explain what your expectation is?
> 
> As pointed out [here](https://github.com/kubernetes/kubernetes/issues/122122#issuecomment-1839268455) this is WAI for the current design that graceful node shutdown leaves the terminated pods around. I will leave the decision about the design of the feature to the node team.
> 
> For some higher-level controllers, such as StatefulSet, the terminated pods need to be recreated, and to my knowledge all bugs related to this are now fixed (see [here](https://github.com/kubernetes/kubernetes/issues/122122#issuecomment-1846930001)).
> 
> Also, as pointed out [here](https://github.com/kubernetes/kubernetes/issues/122122#issuecomment-1839159729) the accumulated pods would eventually be collected by PodGC, when their number exceeds 12500 (default value for `--terminated-pod-gc-threshold`).

Just to clarify the current status of the issue.
For StatefulSets:
- all possible fixes have been implemented as of 1.29

For the deployment part:
- The completed pods will not be GCed automatically until it hits the count hits `--terminated-pod-gc-threshold` so they stay there as intended.
-  Is the timing for the deployment pod restart/reschedule problematic??

**Comment by mimowo** [comment]
> * all possible fixes have been implemented as of 1.29

Right, the fix https://github.com/kubernetes/kubernetes/pull/121389 was also cherry-picked to [1.27.9](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.27.md#changelog-since-v1278) and [1.28.5](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.28.md#changelog-since-v1284).

> The completed pods will not be GCed automatically until it hits the count hits --terminated-pod-gc-threshold so they stay there as intended.

correct

> Is the timing for the deployment pod restart/reschedule problematic??

Do you mean the fact that the deletion of the terminal pods is delayed in time can affect recreation of pods in Deployments? AFAIK there are no open issues related to that.

**Comment by tuibeovince** [comment]
Another StatefulSet concern:
> For some higher-level controllers, such as StatefulSet, the terminated pods need to be recreated, and to my knowledge all bugs related to this are now fixed (see [here](https://github.com/kubernetes/kubernetes/issues/122122#issuecomment-1846930001)).

To confirm; the fix in statefulsets were tested in two environments. While we were able to confirm that in a single node cluster, the terminated pods in a node are indeed recreated when the node is rebooted; multiple node clusters (in our case 2 nodes) first have the pods in shutting down node rescheduled to the still Ready node. But the pods do not get scheduled back into the rebooted node after becoming Ready again. 

Is this behavior according to what was expected of the fix?

**Comment by tuibeovince** [comment]
Supplementary info:

I know that the fix for StatefulSets has already been implemented but I wanted to share a peculiar behavior I observed in v1.27.8 and v1.28.4.

I made a 2 worker node, 1 control-plane node cluster and used the same stateful set config from https://github.com/kubernetes/kubernetes/issues/118310#issuecomment-1698823460

As I was trying out a number of experimentations with the StatefulSet pods, the "Completed" pods do not get cleaned up, as reported in the issue, if the node was "rebooted" with `systemctl reboot`

However, just powering off the node (i.e. using `systemctl poweroff`) and not turning it on again for a while, resulted to having the "Completed" pods eventually rescheduled to a `Ready` node after over 5 minutes and ended up `Running`

```
# kubectl get no,po -o wide

NAME                             STATUS     ROLES           AGE    VERSION   INTERNAL-IP       EXTERNAL-IP   OS-IMAGE          KERNEL-VERSION          CONTAINER-RUNTIME

node/localhost.localdomain       Ready      control-plane   107s   v1.27.8   XXXX              <none>        CentOS Stream 8   4.18.0-544.el8.x86_64   containerd://1.6.28

node/localmachine2.localdomain   Ready      <none>          75s    v1.27.8   YYYY              <none>        CentOS Stream 8   4.18.0-544.el8.x86_64   containerd://1.6.28

node/localmachine3.localdomain   NotReady   <none>          72s    v1.27.8   ZZZZ              <none>        CentOS Stream 8   4.18.0-544.el8.x86_64   containerd://1.6.28

NAME            READY   STATUS      RESTARTS   AGE   IP            NODE                        NOMINATED NODE   READINESS GATES

pod/example-0   1/1     Running     0          59s   AAAA              localmachine2.localdomain   <none>           <none>

pod/example-1   0/1     Completed   0          53s   BBBB              localmachine3.localdomain   <none>           <none>

pod/example-2   1/1     Running     0          48s   CCCC              localmachine2.localdomain   <none>           <none>

pod/example-3   0/1     Completed   0          42s   DDDD              localmachine3.localdomain   <none>           <none>
```
and after a little over 5 minutes: 
```
# kubectl get no,po -o wide

NAME                             STATUS     ROLES           AGE     VERSION   INTERNAL-IP       EXTERNAL-IP   OS-IMAGE          KERNEL-VERSION          CONTAINER-RUNTIME

node/localhost.localdomain       Ready      control-plane   6m54s   v1.27.8  XXXX              <none>        CentOS Stream 8   4.18.0-544.el8.x86_64   containerd://1.6.28

node/localmachine2.localdomain   Ready      <none>          6m22s   v1.27.8   YYYY              <none>        CentOS Stream 8   4.18.0-544.el8.x86_64   containerd://1.6.28

node/localmachine3.localdomain   NotReady   <none>          6m19s   v1.27.8   ZZZZ              <none>        CentOS Stream 8   4.18.0-544.el8.x86_64   containerd://1.6.28

NAME            READY   STATUS    RESTARTS   AGE     IP            NODE                        NOMINATED NODE   READINESS GATES

pod/example-0   1/1     Running   0          6m6s    AAAA              localmachine2.localdomain   <none>           <none>

pod/example-1   1/1     Running   0          12s     BBBB              localmachine2.localdomain   <none>           <none>

pod/example-2   1/1     Running   0          5m55s   CCCC              localmachine2.localdomain   <none>           <none>

pod/example-3   1/1     Running   0          7s      DDDD              localmachine2.localdomain   <none>           <none>
```

**Comment by tuibeovince** [comment]
@mimowo @SergeyKanzhelev 

I was wondering who gets to decide or how will it be decided if this issue should be closed. I believe closing this would be a good step towards the GA of this GracefulNodeShutdown.

As far as the initial issue is concerned;

- "Never-cleaned-up pods" from Deployment are designed to be removed when the PodGC's threshold is met; this is working as designed. Deployment pods which aren't cleaned up theoretically does not impact the workflow since they are created as new pods.
- The suggestion to implement a way for Deployment controllers to clean up the finished pods was left hanging, but apparently it has no urgency of implementation
- The StatefulSet problem has been mitigated by implementing the restart/reschedule of Completed pods in 1.29.0 and backported in 1.28.5 and 1.27.9.

From my perspective, there isn't anything left to settle regarding these. Is this correct?

**Comment by tuibeovince** [comment]
Would it be possible to consider this issue resolved?

To wrap up the last few comments I made in this issue in the past months:

Investigating the StatefulSet fix implemented in https://github.com/kubernetes/kubernetes/pull/121389
I have found that prior to the fix implementation of https://github.com/kubernetes/kubernetes/pull/121389, StatefulSet pods has a case where they get cleaned up,
and that is when the node that was shutdown does not get started again before the pods timeout in a NotReady Node (as described in https://github.com/kubernetes/kubernetes/issues/122122#issuecomment-2051376319)
```
+---------------------------------------------+--------------------------------------------+--------------------------------------------------------------------------+
|                      /                      |              Pre #121389 fix               |                             Post #121389 fix                             |
+---------------------------------------------+--------------------------------------------+--------------------------------------------------------------------------+
| (1) Node shutdown ONLY                      | StatefulSet pods get cleaned after timeout | StatefulSet pods get cleaned immediately, rescheduled to a Ready Node.   |
| (2) Node restart immediately/before timeout | StatefulSet pods stay uncleaned            | StatefulSet pods get cleaned immediately, rescheduled to a Ready Node.   |
| (3) Node shutdown and started after timeout | StatefulSet pods get cleaned after timeout | StatefulSet pods get cleaned immediately, rescheduled to a Ready Node.   |
+---------------------------------------------+--------------------------------------------+--------------------------------------------------------------------------+
- The timeout here is 5 minutes
- Node shutdowns are triggered by `systemctl poweroff`; restarts are triggered by `systemctl reboot`
- (1) and (3) seemed to have an already passable GracefulNodeShutdown behavior, but the immediate rescheduling to the Ready Node is generally an improvement.
- (2) displays a behavior same as this Issue's reported uncleaning, and is now fixed.
```

With the following assumptions:
- Interaction of GracefulNodeShutdown and Deployments are working as intended and no further action is needed;
- StatefulSet fixes implemented in https://github.com/kubernetes/kubernetes/pull/121389 are resolved the never-cleaned pods problem
the issue as originally reported could be resolved by using k8s versions implementing #121389 ; and
- The issue was covering only the pods from Deployment and StatefulSets, other potential problems in other controllers are out of scope and/or are tracked in a different issue

I think nothing else can be done for this issue, thus can be considered somehow to be resolved.

Thank you

**Comment by buroa** [comment]
Closing this issue in leu for #116965
